src/app/qgshandlebadlayers.cpp

```diff
@@ -378,20 +370,69 @@ void QgsHandleBadLayers::apply()
     QTableWidgetItem *item = mLayerList->item( i, 4 );
     QString datasource = item->text();
 
-    node.namedItem( QStringLiteral( "datasource" ) ).toElement().firstChild().toText().setData( datasource );
-    if ( QgsProject::instance()->readLayer( node ) )
+    bool dataSourceChanged { false };
+    const QString layerId { node.namedItem( QStringLiteral( "id" ) ).toElement().text() };
+    const QString provider { node.namedItem( QStringLiteral( "provider" ) ).toElement().text() };
+    const QString name { mLayerList->item( i, 0 )->text() };
+
+    // Try first to change the datasource of the existing layers, this will
+    // maintain the current status (checked/unchecked) and group
+    if ( QgsProject::instance()->mapLayer( layerId ) )
+    {
+      QgsDataProvider::ProviderOptions options;
+      QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( QgsProject::instance()->mapLayer( layerId ) );
+      if ( vl )
+      {
+        vl->setDataSource( datasource, name, provider, options );
+        dataSourceChanged = vl->isValid();
+      }
+      else
+      {
+        QgsRasterLayer *rl = qobject_cast<QgsRasterLayer *>( QgsProject::instance()->mapLayer( layerId ) );
+        if ( vl )
```

Comment:
```suggestion
        if ( rl )
```

... or that's never gonna happen